File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        final List<Integer> ret = new ArrayList<>();
        final Deque<TreeNode> stack = new ArrayDeque<>();

        for (; ; ) {
            for (; root != null; root = root.left) stack.push(root);

            if (stack.isEmpty()) break;

            root = stack.pop();
            ret.add(root.val);
            root = root.right;
        }

        return ret;
    }
}

File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ret;
        stack<TreeNode*> nodes;
        goto goLeft;

        while (!nodes.empty()) {
            root = nodes.top();
            nodes.pop();
            ret.push_back(root->val);
            root = root->right;
    goLeft:
            for (; root; root = root->left) nodes.push(root);
        }

        return ret;
    }
};
